### PR TITLE
Add test for input

### LINKED ISSUE #14: Add tests for inputs
It would be good to start adding some non-visual testing. An easy first test set is the input models. Here is a list of possible tests (though more would be welcome):
- Averaging the spike count of `ConstantRate` over a fixed window of time should be `ConstantRate.rate`
- The frequency constructor for `ConstantRate` should match the rate constructor
- A `StepCurrent` should be zero < threshold time and positive >= threshold time
- The spikes of a inhomogeneous Poisson input with a constant `λ` should be distributed as a [Poisson process](https://en.wikipedia.org/wiki/Poisson_point_process)

### PR DESCRIPTION
**Pull request by RashikaKarki**
Fix #14 
@darsnack
- [X] Averaging the spike count of `ConstantRate` over a fixed window of time should be `ConstantRate.rate`
- [X] The frequency constructor for `ConstantRate` should match the rate constructor
- [X] A `StepCurrent` should be zero < threshold time and positive >= threshold time
- [ ] The spikes of a inhomogeneous Poisson input with a constant λ should be distributed as a Poisson process

I am a bit confused on:
>  positive >= threshold time

was this supposed to be
>  positive > threshold time

### CODE REVIEW

File: test/input-test.jl [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        for i in 1:T
+            spikes = excite!(srm, input, T)
+            push!(spikescount, length(spikes))
+        end
+
+        avgspikecount = mean(spikescount)/T
+
+        @test Float32(round(avgspikecount,digits = 2)) == input.rate
+        
+    end
```
Comment (darsnack) [inline test/input-test.jl]: You don't need to use any neurons like LIF or SRM0 for this test. I was thinking of something much simpler like:
```julia
@testset "ConstantRate" begin
    T = 10_000
    rate = rand()
    input = ConstantRate(rate)
    @test count(x -> x > 0, [input(t) for t in 1:T]) / T ≈ rate
end
```

File: test/input-test.jl [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+   # Assert frequency constructor matches the rate constructor
+   @testset "ConstantRate Constructor" begin
+
+    freq = 0.05
+    dt = 1.0
+    rate = freq * dt
+
+    @test ConstantRate(freq, dt).rate == ConstantRate(rate).rate
+
+    end
```
Comment (darsnack) [inline test/input-test.jl]: Can you also add a test for `dt != 1`?

File: test/input-test.jl [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    evaluate!(stepcurrent, t) != 0 ? flag = false : flag = true
+                        
+                else
+                    (evaluate!(stepcurrent, t) > 0) != true ? flag = false : flag = true
+                end
+            end     
+
+            @test flag == true
+        end
+    end
```
Comment (darsnack) [inline test/input-test.jl]: This test can be simplified.
```julia
τ = 10
T = 25
stepcurrent = StepCurrent(τ)
prethresh = [stepcurrent(t) for t in 1:(τ - 1)]
postthresh = [stepcurrent(t) for t in τ:T]

@test all(x -> x == 0, prethresh)
@test all(x -> x > 0, postthresh)
```

File: test/input-test.jl [hunk 4]
```diff
@@ -0,0 +1,96 @@
+using SpikingNN
+using Statistics
+using Test
+
+@testset "input" begin
```
Comment (darsnack) [inline test/input-test.jl]: Can you rename the test set as "Inputs"

**Review by darsnack — CHANGES_REQUESTED** [review]
Good job! I requested some changes to simplify some of the tests. You are correct about the threshold, hopefully my comment on that test makes clear what's correct.

Also, can you generally fix the indentation and get rid of extra whitespace? It's fine to have whitespace between test sets, but don't add whitespace after every line. The indentation should be four spaces.

File: test/runtests.jl [hunk 5]
```diff
@@ -23,9 +23,13 @@ testfiles = [
     "stdp-test.jl"
 ]
 
-@testset "SpikingNN.jl" begin
+@testset "VisualRegressionTests.jl" begin
```
Comment (darsnack) [inline test/runtests.jl]: I’d prefer more human readable names. Here “Visual Regression Tests”.

File: test/runtests.jl [hunk 6]
```diff
@@ -23,9 +23,13 @@ testfiles = [
...
     ENV["GKSwstype"] = "100"
     for testfile in testfiles
         include(testfile)
     end
+end
+
+@testset "inputs.jl" begin
```
Comment (darsnack) [inline test/runtests.jl]: Here “Input Models”.

File: test/input-test.jl (hunk 3 above)
Comment (Diaga) [reply test/input-test.jl]: @darsnack This still seems inconsistent with Step Current current definition which is returning τ if > threshold only whereas this test case checks for τ >= threshold.
Should the actual following __evaluate!__ method be changed?
```
evaluate!(input::StepCurrent, t::Integer; dt::Real = 1.0) = (t * dt > input.τ) ? t : zero(t)
```

File: test/input-test.jl [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+@testset "ConstantRate" begin
+    T = 10_000
+    rate = rand()
+    input = ConstantRate(rate)
+    @test isapprox(count(x -> x > 0, [input(t) for t in 1:T]) / T, rate, atol=0.005)
+end
+
+# Assert frequency constructor matches the rate constructor
+@testset "ConstantRate Constructors" begin
+    freq, dt = rand(), rand()
```
Comment (darsnack) [inline test/input-test.jl]: This won’t consistently work. Frequency is in Hertz. If `freq == 0.01`, then you need to enforce `dt <= 0.01`.

**Review by darsnack — CHANGES_REQUESTED** [review]
Thanks. Just want to check, are the two of you working together? I made this issue for @RashikaKarki to tackle and I don’t know how MLH is organizing fellows to work on issues.

File: test/input-test.jl (hunk 3 above)
Comment (darsnack) [reply test/input-test.jl]: Ugh you’re right. Functionally it doesn’t matter which side the equal is on. We should change the test to match `evaluate!`’s behavior.

**Comment by Diaga** [comment]
Yes, we are divided in teams of 2 in a pod of 10-12 members.
However, if you feel we should each tackle a separate issue, let's do that.

**Comment by darsnack** [comment]
No it’s totally cool. Please divide up tasks how you see fit. Just wanted to double check.

File: test/input-test.jl (hunk 1 above)
Comment (RashikaKarki) [reply test/input-test.jl]: Thanks. Wasn't aware of this. This sure is way too cleaner.

File: test/input-test.jl (hunk 7 above)
Comment (RashikaKarki) [reply test/input-test.jl]: I did add a condition but I am a bit confused about why we need this condition. Assuming that ` freq = 0.01` and `dt = 0.5` which makes `rate = 0.005` but given this condition the test does pass. So, I think I am missing something here and I am not sure if the way I dealt with this is correct.

File: test/input-test.jl [hunk 8]
```diff
@@ -8,7 +8,8 @@ end
 
 # Assert frequency constructor matches the rate constructor
 @testset "ConstantRate Constructors" begin
-    freq, dt = rand(), rand()
+    freq = rand()
+    freq == 0.01 ? dt = rand() * 0.01 : dt = rand()
```
Comment (darsnack) [inline test/input-test.jl]: I messed up my example. Let me be more clear. `freq` is in Hertz. So if `freq == 20`, then that means "fire 20 times every second" or "fire once every 0.05 seconds". So, in general, if `dt` must be random between `[0, 1 / freq]`. In order to "fire once every X seconds" we must simulate at a resolution of X or smaller.

Probably, we want `freq` to be a random number between `[1, 100]` and `dt` between `(0, 1 / freq]`.

Also don't use ternary operators for assignments like that. It's hard to read. The preferred style would be
```julia
destination = condition ? trueval : falseval
```

File: test/input-test.jl [hunk 9]
```diff
@@ -8,7 +8,8 @@ end
 
 # Assert frequency constructor matches the rate constructor
 @testset "ConstantRate Constructors" begin
```
Comment (darsnack) [inline test/input-test.jl]: We can group this with the previous test set into a single "ConstantRate" set that does both tests one after the other.

**Review by darsnack — CHANGES_REQUESTED** [review]
Some more small changes.

**Review by darsnack — COMMENTED** [review]
Looks great! Once the Poisson test is complete, we can merge.

**Comment by darsnack** [comment]
Okay I am merging this as is, since the Poisson input test seems to be more complicated than expected. The fix may be something I need to do, but if it turns out to be a good issue, then I will open it for both of you to work on.
